Email editor - color palette support [MAILPOET-5741]

costasovo opened the pull request:
## Description

This PR adds support for colors selected from the color palette. Prior to this change, values from the palette were ignored by the renderer.
As a bonus, I also added support for the site's theme palette (tested with TT4)

![Screenshot 2024-02-22 at 15 45 17](https://github.com/mailpoet/mailpoet/assets/1082140/e7b587ee-a26e-488c-bf5d-a8a2e2d67125)


## Code review notes

This doesn't cover the color palette support for border colors. I created an extra ticket for that.

## QA notes

_N/A_

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5741]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5741]: https://mailpoet.atlassian.net/browse/MAILPOET-5741?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Add color palette classes to CSS for renderer

[MAILPOET-5741]

costasovo committed:
Adjust Paragraph block to work with inlined color palette CSS

[MAILPOET-5741]

costasovo committed:
Adjust Paragraph block to work with inlined color palette CSS

[MAILPOET-5741]

costasovo committed:
Adjust the list block renderer to support inlined colors

[MAILPOET-5741]

costasovo committed:
Add support for palette colors for column and columns

[MAILPOET-5741]

costasovo committed:
Add colors from palette support to the button

[MAILPOET-5741]

costasovo committed:
Add getAttributeValueByTagName to dom helper

I'm adding this method to cover use-cases where we need to extract classes
from a block's HTML.
[MAILPOET-5741]

costasovo committed:
Use dom helper for extracting classes from block HTML

[MAILPOET-5741]

costasovo committed:
Move getElementInnerHTML from button to DomDocumentHelper

[MAILPOET-5741]

costasovo committed:
Extract Theme related methods from SettingsController as ThemeController

I want to avoid the SettingsController becoming some kind of a god class
so in this step, I extracted themejson-related stuff to an extra class.

In the future, we should also move methods for manipulating CSS to a helper.
[MAILPOET-5741]

costasovo committed:
Cache WP_Theme_JSON instance

This is a slight improvement to avoid building WP_Theme_JSON instance repeatedly.
[MAILPOET-5741]

costasovo committed:
Add support for the site theme's color palette

[MAILPOET-5741]

costasovo committed:
Remove unused variable from columns block renderer

[MAILPOET-5741]

costasovo committed:
Remove unused method from SettingsController

[MAILPOET-5741]

lysyjan commented on mailpoet/lib/EmailEditor/Integrations/Core/Renderer/Blocks/Column.php:
```diff
@@ -40,19 +50,22 @@ private function getBlockWrapper(array $parsedBlock, SettingsController $setting
       'width' => $width,
       'vertical-align' => $verticalAlign,
     ];
+    $mainCellClasses = 'email_column_wrap';
```
What is the purpose of this class? I haven't found it. If it's still relevant, a comment would be nice here.

costasovo replied:
Aha. Good catch! It is a leftover after an idea I abandoned. I'll remove it. 👍

lysyjan commented on mailpoet/lib/EmailEditor/Integrations/Core/Renderer/Blocks/Columns.php:
```diff
@@ -44,16 +54,17 @@ private function getBlockWrapper(array $parsedBlock, SettingsController $setting
           <tbody>
             <tr>
-              <td style="font-size:0px;background:' . $backgroundColor . ';background-color:' . $backgroundColor . ';padding-left:' . $paddingLeft . ';padding-right:' . $paddingRight . ';padding-bottom:' . $paddingBottom . ';padding-top:' . $paddingTop . ';text-align:left;">
+              <td style="font-size:0px;padding-left:' . $paddingLeft . ';padding-right:' . $paddingRight . ';padding-bottom:' . $paddingBottom . ';padding-top:' . $paddingTop . ';text-align:left;">
```
You used the `esc_attr` function in the previous class. I a bit wonder why it's not also used here. Maybe we can create a ticket to unify all renderers in this way. What do you think?

costasovo replied:
I addressed [this particular line in a follow-up PR](https://github.com/mailpoet/mailpoet/pull/5450/commits/caceefe0bac052c791c28dcc7210a8767adc06e8#diff-dcddd58f27adb701f13eac70e8f89fc681684e1d45d39df71e77f8d73ca7af23R66) but there are more unescaped places.
@lysyjan I'm glad you asked. I was thinking about adding a ticket for this so I'll prepare a ticket. IMO we should add proper escaping to all the places where we build the HTML.

lysyjan reviewed (dismissed):
Good job! I have only a question about what I don't understand.

costasovo committed:
Remove an unnecessary class from a column wrapper

[MAILPOET-5741]

costasovo commented:
@lysyjan Thank you for the detailed review! I added a commit that fixes the unwanted class. As for the escaping, I created an extra ticket and added you as a reviewer.

lysyjan reviewed (approved):
@costasovo I'm not sure about QA here. I'm assigning it back to you. Just so you know, you can decide about the next step.

costasovo commented:
@lysyjan Thanks. There are still a few known issues I would like to fix before we pass rendering to QA so I'm going to merge this.
